Various QoL MapEditor changes and issue 1856

SJuliez opened the pull request:
- Several MapEd file dialogs (save, load etc.) now, finally, remember their size.
- The validation dialogs now appear centered and react to ESC or Enter.
- The board validation upon loading a board now does show the error report because it can be different from what a later validation shows, especially when buildings have errors.
- The minimap (in the map editor) now shows invalid hexes as bright red.
- The MapEd frame title now shows if the board has been changed (with the usual "*"). A fresh random board does not count as changed.
- If the board is changed, the MapEd will now not just close when the window "X" is pressed but ask if the user wants to save the board first.
- The RandomMap dialog at the startup of the MapEd can now be disabled in the dialog.
- Bridges and other orthos now show up in the work hex - Resolves #1856

Windchild292 commented on megamek/src/megamek/client/ui/swing/BoardEditor.java:
```diff
@@ -1345,16 +1376,19 @@ private void updateWhenSelected() {
+            RandomMapDialog rmd = new RandomMapDialog(frame, this, null, mapSettings);
+            userCancel = rmd.activateDialog(bv.getTilesetManager().getThemes());
+        }
+        if (!userCancel) {
+            board = BoardUtilities.generateRandom(mapSettings);
+            game.setBoard(board);
+            curfile = null;
+            butSourceFile.setEnabled(false);
+            frame.setTitle(Messages.getString("BoardEditor.title")); //$NON-NLS-1$
+            menuBar.setBoard(true);
     		bvc.doLayout();
     		resetUndo();
     		choTheme.setSelectedItem(mapSettings.getTheme());
```
I think this is tab versus space spacing, which would be good to fix.

Windchild292 commented on megamek/src/megamek/client/ui/swing/BoardEditor.java:
```diff
@@ -1759,21 +1807,17 @@ public void actionPerformed(ActionEvent ae) {
                 } catch (IOException e) {
-                    JOptionPane.showMessageDialog(this, "Could not open the file "+curfile+". "+e.getMessage());
+                    ignoreHotKeys = true;
+                    JOptionPane.showMessageDialog(frame, "Could not open the file "+curfile+". "+e.getMessage());
```
```suggestion
                    JOptionPane.showMessageDialog(frame, "Could not open the file " + curfile + ". " + e.getMessage());
```

Windchild292 commented on megamek/src/megamek/client/ui/swing/BoardEditor.java:
```diff
@@ -2069,6 +2124,31 @@ public boolean shouldIgnoreHotKeys() {
+     *  and a "*" if the board has been changed.
+     */
+    private void setFrameTitle() {
+        String title = Messages.getString("BoardEditor.title"); //$NON-NLS-1$
+        if (curfile != null) {
+            title = Messages.getString("BoardEditor.title") + curfile; //$NON-NLS-1$
+        }
```
```suggestion
        String title = Messages.getString("BoardEditor.title"); //$NON-NLS-1$
        if (curfile != null) {
            title += curfile;
        }
```

Windchild292 commented on megamek/src/megamek/client/ui/swing/MiniMap.java:
```diff
@@ -728,6 +728,13 @@ protected synchronized void drawMapOrig() {
                         paintCoord(gg, j, k, true);
                     }
                     addRoadElements(h, j, k);
+                    // Color invalid hexes red when in the Map Editor
+                    if ((m_game != null) && 
+                            (m_game.getPhase() == IGame.Phase.PHASE_UNKNOWN)
+                            && !h.isValid(null)) {
+                        gg.setColor(Color.RED);
+                        paintCoord(gg, j, k, true);
+                    }
```
Is red really the best choice, given that some people can't see it?

Windchild292 commented on megamek/src/megamek/client/ui/swing/RandomMapDialog.java:
```diff
@@ -271,8 +274,22 @@ private JPanel setupTopPanel() {
+        // The main panel with the Okay, Cancel etc. buttons
         JPanel panel = new JPanel(new FlowLayout(FlowLayout.CENTER, 2, 2));
-
+        
```
```suggestion

```

SJuliez commented:
Thanks for checking, @Windchild292 
I corrected some tab spacings, updated the Message system to the more useful (I think) Object... instead of Object[] parameter (can save some "new Object[] {...}"). Also removed a few one-line ifs which were my own wrong style originally. (I just recently noticed there is a coding style guide in the wiki)
Concerning Color.Red and visibility, red is still the usual warning/bad color everywhere, MegaMek or otherwise. I have no problem using GUIPreferences.getWarningColor() or something if it exists but I'd rather not arbitrarily start using Blue now (or whatever would be a good color, I actually don't know even that).

Windchild292 commented on megamek/src/megamek/client/ui/swing/BoardEditor.java:
```diff
@@ -1751,6 +1761,17 @@ private void showSettings() {
         setdlg.setVisible(true);
     }
+    
+    private void freshBoardSettings() {
```
I'm not sure about this name, as it took me a second to realise what it meant.

SJuliez replied:
Do you have a suggestion? Its not necessarily a new board, not necessarily a loaded board. updateUiForFreshBoard...

Windchild292 replied:
refreshBoardSettings? Not super fitting, but better. refreshBoardUI?

Windchild292 commented on megamek/src/megamek/client/ui/Messages.java:
```diff
@@ -1,16 +1,16 @@
- *  for more details.
- */
+/*  
+* MegaMek - Copyright (C) 2020 - The MegaMek Team  
```
Why are you removing the original side of things?

SJuliez replied:
Please explain. Im not remotely an expert on copyright in software applications so I thought why not update it to the current example in the wiki.

Windchild292 replied:
I did that originally, but was told to instead leave the original and add the new line below it when I first started contributing. Hence the question.

sixlettervariables replied:
Basically we can't erase Ben's copyright on his work. Unless this is a total rewrite.

Windchild292 commented:
If we don't have a getWarningColour, it is a good addition for accessibility (~8.5% of the population is red-green colourblind). Maybe just put a getter in that just returns red, and I'll open up a story to add that for accessibility?

SJuliez commented:
Sounds good. I'll do that.

SJuliez commented:
OK, changed the copyright notice and method name and added an accessibility oriented WARNING_COLOR to the GUIP that defaults to red and currently has no UI for changes.

Windchild292 commented on megamek/src/megamek/client/ui/swing/BoardEditor.java:
```diff
@@ -2137,17 +2143,25 @@ public void paintComponent(Graphics g) {
                 }
                 // add level and INVALID if necessary
+                if (guip.getAntiAliasing()) {
+                    ((Graphics2D)g).setRenderingHint(
```
```suggestion
                    ((Graphics2D) g).setRenderingHint(
```
